#905 Check for valid redshifts

jmeyers314 opened the pull request:
Adds some sanity checks and unit tests for valid redshift inputs to `SED.atRedshift`.

rmjarvis commented on galsim/utilities.py:
```diff
@@ -1145,6 +1145,8 @@ def combine_wave_list(*args):
             red_limit = min(red_limit, obj.red_limit)
         wave_list = np.union1d(wave_list, obj.wave_list)
     wave_list = wave_list[(wave_list >= blue_limit) & (wave_list <= red_limit)]
+    if blue_limit > red_limit:
```
Should this maybe check if wave_list is empty?  Or can that only happen if blue_limit > red_limit?  

Also, even only having a single element is likely a problem, so might be worth giving an error for that too.

jmeyers314 replied:
Actually, I did try `if len(wave_list) == 0` before settling on `if blue_limit > red_limit`.  The former led to a dozen or so errors in the unit tests though.  Empty wave_lists are valid if you have an SED or Bandpass with an analytic wavelength dependence instead of one based off a LookupTable, so I don't think that's the right thing to check for.

I agree a single element would be suspicious.  I think the right way to handle that may be to change `>` -> `>=` above, rather than looking at wave_list directly.

rmjarvis commented:
Here's a quick script that might be added to your unit tests somewhere:
```
sed = galsim.SED(galsim.LookupTable([1,2,3,4,5], [1.1,1.9,1.4,1.8,2]),
                 wave_type='nm', flux_type='fphotons')
bp = galsim.Bandpass(galsim.LookupTable([4,5],[1,1], interpolant='linear'),
                     wave_type='nm')

for z in [0, 1, 2, 3, 3.5, 3.99, 4, 4.01]:
    print('z = ',z,' flux = ',sed.atRedshift(z).calculateFlux(bp))
```
The question is really what should happen for the last two redshifts.  Currently this is the output:
```
z =  0  flux =  1.9
z =  1  flux =  1.78470982143
z =  2  flux =  1.63869047619
z =  3  flux =  1.24875837054
z =  3.5  flux =  0.58363707623
z =  3.99  flux =  0.0110121850677
z =  4  flux =  0.0
Traceback (most recent call last):
  File "junk.py", line 10, in <module>
    print('z = ',z,' flux = ',sed.atRedshift(z).calculateFlux(bp))
  File "/Users/Mike/lib/python2.7/site-packages/galsim/sed.py", line 679, in calculateFlux
    x, _, _ = galsim.utilities.combine_wave_list(self, bandpass)
  File "/Users/Mike/lib/python2.7/site-packages/galsim/utilities.py", line 1149, in combine_wave_list
    raise RuntimeError("Empty wave_list intersection.")
RuntimeError: Empty wave_list intersection.
```
I actually think that is probably fine for z=4 where the intersection has 1 element.  The flux comes out 0, which is correct.  

For the last one though, should we allow this without error, and give a flux of 0?  Or do you think the error is the right runtime behavior?  (And if so, maybe something from calculateFlux would be better, since this is still a little obscure for a user to figure out.)

jmeyers314 commented:
>For the last one though, should we allow this without error, and give a flux of 0? Or do you think the error is the right runtime behavior?

We raise an exception when requesting the flux density of an SED outside of its defined range, so I think the right behavior is is to raise an exception for calculateFlux when that function would have asked about an out-of-bounds wavelength.  I agree that it would be helpful for that message to come from calculateFlux itself though.

(Note that the opposite behavior is true for Bandpass though, where requesting the throughput for an out-of-bounds wavelength does return 0.0.  My intuition for both of these choices was that users probably don't mean that an object's SED is really zero outside the range they specify, it's just unspecified.  But they probably do mean that a Bandpass throughput is zero outside of the specified range.)

rmjarvis commented:
> users probably don't mean that an object's SED is really zero outside the range they specify,

I think I agree with that.  But in that case, I guess there should be an error thrown if any part of the bandpass is outside the sed defined range.  So z=3.5 should raise an exception too in my example above.  (Also 3.01.)

jmeyers314 commented:
>I guess there should be an error thrown if any part of the bandpass is outside the sed defined range.

Yep.  I agree with that.  Will fix...

jmeyers314 commented:
This is making me think we should also remove the bolometric flux calculation, which should nominally require the SED to be defined all the way from 0 -> infinity (so only possible using an analytic function SED, not a LookupTable.)  It's pretty easy to roll a new bandpass with throughput 1 and blue/red limits equal to that of the SED if anyone actually does want/use this.

rmjarvis commented:
> we should also remove the bolometric flux calculation

I don't have any preference.  Probably not many people use it.  But if we, at this point, it would just be a deprecation, so we can see if anyone complains...

jmeyers314 commented:
Okay, I think I fixed the invalid evaluation of SEDs inside `calculateFlux` bug, and also deprecated the bolometric flux.

I also simplified the default values of `SED.blue_limit` and `SED.red_limit` to be 0 and infinity instead of `None`.  This eliminated a handful of `if blue_limit is not None` kinds of checks that are no longer necessary.

rmjarvis commented on galsim/sed.py:
```diff
@@ -230,21 +229,7 @@ def dimensionless(self):  # for convenience
-                red_limit = min([red_limit, other.red_limit])
-
-        return blue_limit, red_limit
+        return max([self.blue_limit, other.blue_limit]), min([self.red_limit, other.red_limit])
```
Apparently, this function is not actually used by anything, so could just remove it.

rmjarvis commented on galsim/sed.py:
```diff
@@ -651,29 +632,26 @@ def calculateFlux(self, bandpass):
+                 "If you need this functionality, you can use a pseudo-bolometric Bandpass created "
+                 "with:  bp = Bandpass('1', blue_limit=sed.blue_limit, red_limit=sed.red_limit)")
+            bp = galsim.Bandpass('1', self.blue_limit, self.red_limit)
+            return self.calculateFlux(bp)
```
This needs a test in test_deprecated.  Normally the existing test of this functionality would be moved over to there with the check_dep call.  But there wasn't one, so I guess you'll need to make something for this functionality.  Probably could make a simple sed that has an analytic bolometric flux and make sure it comes out right.

rmjarvis replied:
Looks like the depr message should get the 'nm' edit too...

jmeyers314 commented:
Thanks @rmjarvis.  I think I fixed the missing test coverages.  Let me know if you see anything else missing.  Otherwise, I'll plan to merge tomorrow.
